snap/squashfs2/squashfs2.go

```diff
@@ -0,0 +1,531 @@
+	default:
+		inodeSize := getDefaultInodeSize(inodeType)
+		if inodeSize == 0 {
+			return nil, fmt.Errorf("squashfs: invalid inode type %d", inodeType)
+		}
+
+		inodeBuffer := make([]byte, inodeSize+2)
+		copy(inodeBuffer, typeBuffer)
+		err = sfs.inodeReader.read(inodeBuffer[2:])
+		if err != nil {
+			return nil, err
+		}
+		return inodeBuffer, nil
+	}
+}
+
+func (sfs *SquashFileSystem) loadRootDirectory() error {
+	inodeBuffer, err := sfs.readInodeData()
+	if err != nil {
+		return err
+	}
+
+	inode := &squashfs_inode_dir{}
+	inode.parse(inodeBuffer)
+
+	sfs.rootDirectory = sfs.directoryCreate(inode)
+	return nil
+}
+
+func (sfs *SquashFileSystem) Close() error {
+	return sfs.stream.Close()
+}
+
+func (sfs *SquashFileSystem) readDirectoryEntryInode(entry *squashfs_dir_entry) ([]byte, error) {
+	if err := sfs.inodeReader.seek(int64(entry.startBlock), int(entry.offset)); err != nil {
+		return nil, err
+	}
+	return sfs.readInodeData()
+}
+
+func (sfs *SquashFileSystem) createDirectoryFromDirectoryEntry(entry *squashfs_dir_entry) (*directory, error) {
+	if !entry.isDirectory() {
+		return nil, fmt.Errorf("squashfs: %s is not a directory", entry.name)
+	}
+
+	inodeBuffer, err := sfs.readDirectoryEntryInode(entry)
+	if err != nil {
+		return nil, err
+	}
+
+	inode := &squashfs_inode_dir{}
+	inode.parse(inodeBuffer)
+	return sfs.directoryCreate(inode), nil
+}
+
+func (sfs *SquashFileSystem) readFileFromDirectoryEntry(entry *squashfs_dir_entry) ([]byte, error) {
+	if entry.isDirectory() {
+		return nil, fmt.Errorf("squashfs: %s is must not be a directory", entry.name)
+	}
+
+	if entry.isSymlink() {
+		return nil, fmt.Errorf("squashfs: %s is a symlink, and we do not support this yet", entry.name)
+	}
+
+	inodeBuffer, err := sfs.readDirectoryEntryInode(entry)
+	if err != nil {
+		return nil, err
+	}
+
+	if entry.isRegularFile() {
+		inode := &squashfs_inode_reg{}
+		inode.parse(inodeBuffer)
+		return inode.read_data(sfs)
+	} else {
+		return nil, fmt.Errorf("squashfs: %s is not a regular file", entry.name)
+	}
+}
+
+func (sfs *SquashFileSystem) ReadFile(path string) ([]byte, error) {
+	currentDirectory := sfs.rootDirectory
+
+	// split the provided path into tokens based on '/'
+	tokens := strings.Split(path, "/")
+	for i, token := range tokens {
+		entry, err := currentDirectory.lookupDirectoryEntry(token)
+		if err != nil {
+			return nil, err
+		}
+
+		if i == len(tokens)-1 {
+			// last token, entry shall not be a directory
+			if entry.isDirectory() {
+				return nil, fmt.Errorf("squashfs: %s is a directory", path)
+			}
+			return sfs.readFileFromDirectoryEntry(entry)
+		}
+
+		// otherwise we have to descend into the directory
+		// make sure that is a directory
+		if !entry.isDirectory() {
+			return nil, fmt.Errorf("squashfs: %s is not a directory", path)
+		}
+		currentDirectory, err = sfs.createDirectoryFromDirectoryEntry(entry)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nil, fmt.Errorf("squashfs: %s not found", path)
```

Comment:
I wonder if we should have something like a custom error type, e.g. squashfs.NotExistsError for this case but that can be a followup of course.

Reply:
Really good point! I'll keep this open because this would be really nice to have.